model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion.java

```diff
@@ -0,0 +1,50 @@
+package org.keycloak.connections.jpa.updater.liquibase.custom;
+
+import liquibase.exception.CustomChangeException;
+import liquibase.statement.core.DeleteStatement;
+import liquibase.structure.core.Column;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+public class JpaUpdate26_2_5_RemoveDuplicateMigrationModelVersion extends CustomKeycloakTask {
+
+    @Override
+    protected void generateStatementsImpl() throws CustomChangeException {
+        Set<String> idsToDelete = new HashSet<>();
+
+        String tableName = getTableName("MIGRATION_MODEL");
+        String colId = database.correctObjectName("ID", Column.class);
+
+        try (PreparedStatement ps = connection.prepareStatement(
+                "SELECT ID FROM MIGRATION_MODEL m1 WHERE EXISTS (SELECT 1 FROM MIGRATION_MODEL m2 WHERE m2.VERSION = m1.VERSION AND m2.UPDATE_TIME > m1.UPDATE_TIME)"
+        )) {
+            ResultSet resultSet = ps.executeQuery();
+            while (resultSet.next()) {
+                idsToDelete.add(resultSet.getString(1));
+            }
+        } catch (Exception e) {
+            throw new CustomChangeException(getTaskId() + ": Failed to detect duplicate MIGRATION_MODEL rows", e);
+        }
+
+        AtomicInteger i = new AtomicInteger();
+        idsToDelete.stream()
+                .collect(Collectors.groupingByConcurrent(id -> i.getAndIncrement() / 20, Collectors.toList())) // Split into chunks of at most 20 items
+                .values().stream()
+                .map(ids -> new DeleteStatement(null, null, tableName)
```

Comment:
Is it safe to assume that Liquibase will fill the schema and catalog name with proper defaults if we pass null here? I see that we pass null in other places already.

Reply:
As you mentioned, we use it on other places, so I assume it worked. But I rather explicitly provided it :)